src/librustc_mir/transform/deaggregator.rs

```diff
@@ -0,0 +1,115 @@
+                let src_info = bb.statements[ai.statement_index].source_info;
+                bb.statements.remove(ai.statement_index);
+                for (i, (op, ty)) in ai.operands.iter().zip(ai.types.iter()).enumerate() {
+                    let rhs = Rvalue::Use(op.clone());
+                    let lhs_cast = ai.lhs.clone();
+                    // let lhs_cast = if ai.adt_def.variants.len() > 1 {
+                    //     Lvalue::Projection(Box::new(LvalueProjection {
+                    //         base: ai.lhs.clone(),
+                    //         elem: ProjectionElem::Downcast(ai.adt_def, ai.variant),
+                    //     }))
+                    // } else {
+                    //     ai.lhs.clone()
+                    // };
+                    let lhs_proj = Lvalue::Projection(Box::new(LvalueProjection {
+                        base: lhs_cast,
+                        elem: ProjectionElem::Field(Field::new(i), ty),
+                    }));
+                    let new_statement = Statement {
+                        source_info: src_info,
+                        kind: StatementKind::Assign(lhs_proj, rhs),
+                    };
+                    debug!("inserting: {:?} @ {:?}", new_statement, ai.statement_index + i);
+                    bb.statements.insert(ai.statement_index + i, new_statement);
+                }
+            }
+        }
+    }
+}
+
+fn get_aggregate_statement<'a, 'tcx, 'b>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                         statements: &Vec<Statement<'tcx>>)
+                                         -> Option<AggInfo<'tcx>> {
+    for (i, statement) in statements.iter().enumerate() {
+        //debug!("looking at stmt: {:?}", statement);
+        let StatementKind::Assign(ref lhs, ref rhs) = statement.kind;
+        if let &Rvalue::Aggregate(ref kind, ref operands) = rhs {
+            if let &AggregateKind::Adt(adt_def, variant, substs) = kind {
+                if operands.len() > 0 { // don't deaggregate ()
+                    if adt_def.variants.len() > 1 {
+                        // only deaggrate structs for now
+                        continue;
+                    }
+                    debug!("getting variant {:?}", variant);
+                    debug!("for adt_def {:?}", adt_def);
+                    let variant_def = &adt_def.variants[variant];
+                    if variant_def.kind == VariantKind::Struct {
+                        let types: Vec<_> = variant_def.fields.iter().map(|f| {
+                            f.ty(tcx, substs)
+                        }).collect();
+                        debug!("found a aggregate: {:?}", statement);
+                        debug!("ops: {:?}", operands);
+                        return Some( AggInfo {
+                            statement_index: i,
+                            lhs: lhs.clone(),
+                            operands: operands.clone(),
```

Comment:
Seems weird to me that you’re doing cloning here and then simply discarding the original data on line 41. Why don’t you just return the index of the statement here and then figure out the details like `lhs` and stuff from the statement you remove there (since `Vec::<T>::remove -> T` anyway)?